### PR TITLE
fix(sns): allow tokens to be used in UrlSubscription

### PR DESCRIPTION
**Pull request by spg**
Allows to pass in a dynamic reference as the `url`, like so:

```typescript
myTopic.addSubscription(
  new UrlSubscription("{{resolve:secretsmanager:some-url}}", sns.SubscriptionProtocol.Https)
);
```

without having the CDK complain:
```
Error: URL must start with either http:// or https://
    at new UrlSubscription (/Users/spgingras/myProject/cdk-app/node_modules/@aws-cdk/aws-sns-subscriptions/lib/url.ts:28:13)
...
```

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts [hunk 1]
```diff
@@ -23,17 +23,17 @@ export interface UrlSubscriptionProps {
...
-    if (!url.startsWith('http://') && !url.startsWith('https://')) {
+  constructor(private readonly url: string, private readonly protocol: sns.SubscriptionProtocol, private readonly props: UrlSubscriptionProps = {}) {
+    if (!Token.isUnresolved(url) && !url.startsWith('http://') && !url.startsWith('https://')) {
       throw new Error('URL must start with either http:// or https://');
     }
   }
 
   public bind(scope: Construct, topic: sns.ITopic): void {
-    new sns.Subscription(scope, this.url, {
+    new sns.Subscription(scope, topic.node.uniqueId + 'Url', {
```
Comment (spg) [inline packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts]: This avoids an error:
```
Error: Cannot use tokens in construct ID: ${Token[TOKEN.11]}
```

Comment (rix0rrr) [reply packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts]: Ah, so it does. But now you can only add one URL subscription for any given topic.

Can you use the following expression instead?

```ts
Token.isUnresolved(this.url) ? 'UnresolvedUrl' : this.url
```

So we use the literal one if available, and only fall back to the reuse-breaking identifier if the URL is unresolved?

File: packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts [hunk 2]
```diff
@@ -23,17 +23,17 @@ export interface UrlSubscriptionProps extends SubscriptionProps {
  * @see https://docs.aws.amazon.com/sns/latest/dg/sns-http-https-endpoint-as-subscriber.html
  */
 export class UrlSubscription implements sns.ITopicSubscription {
-  constructor(private readonly url: string, private readonly props: UrlSubscriptionProps = {}) {
-    if (!url.startsWith('http://') && !url.startsWith('https://')) {
+  constructor(private readonly url: string, private readonly protocol: sns.SubscriptionProtocol, private readonly props: UrlSubscriptionProps = {}) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts]: I'd rather have the protocol optionally inside the props, and derive it from the literal URL (if available).

Only force passing it if the URL is unresolved.

Comment (spg) [reply packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts]: @rix0rrr fixed in d524d12

**Comment by spg** [comment]
@rix0rrr not sure what is happening with CodeBuild

File: packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts [hunk 3]
```diff
@@ -23,17 +31,31 @@ export interface UrlSubscriptionProps extends SubscriptionProps {
...
+    if (!this.unresolvedUrl && !url.startsWith('http://') && !url.startsWith('https://')) {
       throw new Error('URL must start with either http:// or https://');
     }
+
+    if (this.unresolvedUrl && props.protocol === undefined) {
+      throw new Error('Must provide protocol if url is unresolved');
+    }
+
+    if (this.unresolvedUrl) {
+      this.protocol = props.protocol!;
```
Comment (eladb) [inline packages/@aws-cdk/aws-sns-subscriptions/lib/url.ts]: The “!” shouldn’t be required here. Try to use !props.protocol above

**Pull request merged**